### ISSUE TITLE
Using OCR for Opening Hours

### ISSUE THREAD

**Issue by ciconi** [westnordost/StreetComplete#1186]
**Use case**
Adding complicated opening hours is often very time consuming, despite the input mask is quite clever. As a user who does street view while walking around the city I would prefer a quicker way. 

**Proposed Solution**
Probably the best solution would be using OCR: The user takes a picture of the opening hours and the app translates it into OSM tags. With the information from the picture the app fills the input mask as a suggestion. Now the user can change fields or save the quest. 
As opening hours are quite simple, it should work quite well. 

There is a ready to use OCR implementation for Android called tesseract. 
https://github.com/tesseract-ocr/

**Comment by westnordost** [comment]
Won't do. First of all, for offline OCR, a machine-learned blob (for every script) is necessary whose size is not negligible, in fact it would be so big that it probably would have to be downloaded once on demand and not packaged with the app. Second, and more importantly, OCR of the text is only the first step. The second step is to recognize and interpret the whole opening hours plate correctly. For a project like this, to develop an AI or machine-learned blob to do this is not achievable at all.
Also note that this would need to work with 100% accuracy if this is not to be used as a prefilled suggestion for the form.

If the opening hours are too complicated, the user can always just leave a note and attach a photo (well, currently not, but will soon be reenabled).

**Comment by rugk** [comment]
[This needs much more pre-processing though](https://github.com/tesseract-ocr/tesseract/wiki/ImproveQuality).
Also what makes you think tesseract is for Android? It's actually a generic C++ module and [Wikipedia only lists](https://en.wikipedia.org/wiki/Tesseract_(software)) desktop OSes as support. Besides of that, it's huge and seems to have specific stuff for supporting each language…

And BTW, I've only ever seen one [FLOSS app](https://github.com/ctodobom/OpenNoteScanner) which that feature, and it uses [this lib](https://opencv.org/) with an awkward setup process, where you have to download the lib (semi)manually. (That thing could be have been changed in the meantime, since I tested, however.)
**Edit:** Actually, it did not do OCR, but just something similar.

Also, even if we'd assume OCR would work flawlessly, we still get some text in a non-standard format, which the app has to parse. As such, we get more or less random formats and especially for "complicated"  cases, this is prone to fail.
We would have to throw AI or so into the mix just for this step, so you'll see, where this goes to…

**Comment by ciconi** [comment]
Thanks for your answers! I expected a little bit that this might be too hard. 
But in regard to tesseract: A android app using it is Mitzuli. https://github.com/artetxem/mitzuli

The foto solution sounds also promising, especially if it is in a way that other users can solve the quest from home!

**Comment by rugk** [comment]
> The foto solution sounds also promising,

Just FYI, because your wording is not clear: That is already implemented. (in the notes)

> users can solve the quest from home

After using SC and attaching pictures of course, yes. 😄

**Comment by okias** [comment]
more than 2 years passed and after today walk with SC, I'd also prefer to have OCR (not have to be perfect, but for most of signs it wouldn't have problem autofill data.

**Comment by rugk** [comment]
Yeah, considering it now again, I see it indeed also a bit differently: AFAIK modern phones do have optimisations/special processor-parts for machine larning things, also we would only need the "evaluation" part in the app – not the learning step of course.

As for training the data: Maybe we could use some generated comments of OSM by this app? Just an idea… otherwise maybe there is some reference data somewhere.

**Comment by ciconi** [comment]
Recently I stumbled over a project that offers OCR functionality for other apps via "Intents", whatever this is. But the test worked quite good, so if it's possible to get that connection running SC only needs to parse the text given back? 

https://github.com/mtotschnig/OCR

I can start collecting pictures of opening hours, if this helps.

**Comment by matkoniecz** [comment]
Even if we would have 100% ideal OCR then the problem of converting OCRed text into OSM opening_hours text remains.

And it is problem that I expect to be horrifically hard, in the class of problems such as self-driving cars.

**Comment by HolgerJeromin** [comment]
> Even if we would have 100% ideal OCR then the problem of converting OCRed text into OSM opening_hours text remains.

Pretty good solved by this project:
https://osm-de.github.io/WebToOSMOH/

**Comment by matkoniecz** [comment]
Fails for first tested string: "Open 10-20 except Monday"

**Comment by starsep** [comment]
OCRing opening hours is a hard problem.
I believe OCR solution such as [Google ML Kit](https://developers.google.com/ml-kit/vision/text-recognition/android)/[Tesseract](https://github.com/tesseract-ocr) would work well enough: not perfect but if human would confirm the prediction I think that could work.
Indeed harder is proper parsing of the text considering different languages, regions, notations, shortcuts etc.

Go Map!! has implemented [OpeningHours OCR](https://github.com/bryceco/GoMap/tree/master/src/iOS/OpeningHours). I haven't tested it as I don't have iOS device.

I think that such a feature would add "wow factor" to StreetComplete if done well but probably is not worth pursuing right now.
CC @kocio-pl as we discussed it,

**Comment by mmann1123** [comment]
Hey all, i just posted a method that is working pretty well using ML Kit text recognition. I am just testing it now, I am sure there's issues I am not considering. See #6679 

https://github.com/user-attachments/assets/ce0adc68-086a-4c64-ac17-94a5cb4e3489

**Comment by westnordost** [comment]
Wow, ... this is a lot. Also, it looks like you poured a lot of work into this, the whole UI and all. It is a shame that you didn't write anything before, you might have saved yourself some effort, then. <strike>Because, most importantly, I assume that you wrote the UI with Android view XMLs, Fragments, Views, right? We are currently migrating to another multiplatform UI framework named Compose, and as a matter of fact, I have been working on recreating the opening hours UI in compose the last few weeks or so, it's almost done. This means, even if I wouldn't have any other complaints about this example, the UI would need to be re-written from scratch again in compose.</strike>

Now, it looks like you use ML Kit text recognition as exactly that, just as OCR. No machine learned opening hours recognition involved as such. This means that the user still needs to, step by step, specify all the elements of the opening hours himself. It's really a shame, as you've put so much work into it already, but I don't think this approach is useful:

1. the user still has to accompany the process all the way. Just these simple opening hours `Mo-Fr 6-15, Sat 6-11, Su off` is about 10 steps and a final confirmation. That's not really less than specifying them manually, plus, more involved. I fear with "custom grouping" etc. it would be even more involved and long-winded.
  Basically, the only thing you do with ML kit text recognition is spare the user from inputting the time with the time picker, that's all. That's not a lot. But you built a lot of UI around it.

2. the input is much less flexible than the manual input. Input is limited to just the most simple opening hours. What about countries where the work week does not start with Monday? What about countries where the work week has a different length? What about shops that close during noon? Well, etc.

3. the last screen seems to be superfluous. Why is there a verification screen if the screen just after that also shows the opening hours?

**Comment by mmann1123** [comment]
No worries I assumed this would be the response, but for me 

1) mobility is the key issue. When using streetcomplete I am always walking somewhere with another purpose -  I don't ever want to stop and stand around. The idea here is that the user can take a photo and walk off.   Agree the OCR might be superfluous, maybe just having a photo with hours dials below it would be just as quick/useful - but honestly just wanted to see if I could get it done with OCR since people pointed out the parsing challenge.  
2) Seems like it would be easy to have common hours combos populated by location. 
3) Yeah that's a good point.

**Comment by westnordost** [comment]
I understand. The opening hours form is one of the few forms so involved that I have to stop during the survey. This bothers me, too. On the other hand, for me, to take a picture and continue to record the data while walking (and looking at the picture) wouldn't solve it for me, because either there will be the next opening hours quest just a few steps later, or another quest. If there is no quest at all, I usually have no issue memorizing easy opening hours like workweek 8-18, saturday 8-12 or whatever for the short time interval between directly staring at the plate and having completed entering the hours.

I've added some features that should make it quicker to input opening hours, such as:
- pre-select workweek (i.e. Mo-Fr in most countries) for the first opening hours days specified
- directly ask weekdays, then times in two subsequent dialogs when adding opening days (i.e. no second tap necessary). Same when adding months
- pre-scroll the time wheel picker to the typical opening hours (9:00-18:00)
- don't show by months and don't offer to add months unless already specified or added via the "Uh..." menu
- when months have been defined, for the second months added, pre-select all the remaining months

In the redone opening hours form, the time input are based on those wheel pickers (partly because compose multiplatform doesn't offer time pickers) instead of the hours dial. Maybe those wheel pickers are faster to use than such a hours dial mainly, as the minute will most often be 00 or 30.

**Comment by mmann1123** [comment]
I would be curious to have some folks try it out you can download the APK here: https://github.com/mmann1123/StreetComplete/actions/runs/20720762765

Also seems like others would like photo option #6480 , would there be interest in designing something where you can work from a photo - skipping the OCR?

**Comment by mnalis** [comment]
> When using streetcomplete I am always walking somewhere with another purpose - I don't ever want to stop and stand around. The idea here is that the **user can take a photo and walk off**

In such cases (when opening hours are quite complex and I'm in a hurry), I usually just choose `Uh...` / `Can't say` / `Leave a note` and add a picture of opening hours (and picture often included other details, like payment cards accepted, phone, website, ref....) in OSM Note; and then actually modify actual OSM elements when I have more time (usually when I get home in JOSM with comfort of big screen and multiple windows, but also Vespucci would do it in a pinch on mobile). 

Another advantage of OSM notes is that often other people are faster and fix that note before I can do it.

Of course, in majority of cases it is easy to just read and remember (one range for Mo-Fr, another for Sa) and solve SC quest from memory while walking towards the next quest.

**Comment by pktiuk** [comment]
Couldn't we just hook OCR with simple LLM to use it to read these opening hours (which could be accepted by user)?       
In most cases it should be able to figure out pieces of text in specific order.

(nowadays more and more of new phones come with some kinds of NPUs)

**Comment by matkoniecz** [comment]
see https://github.com/streetcomplete/StreetComplete/issues/1186#issuecomment-3709210006

**Comment by pktiuk** [comment]
> see [#1186 (comment)](https://github.com/streetcomplete/StreetComplete/issues/1186#issuecomment-3709210006)

I saw it, but I don't think it answers my question. There is nothing about autointerpreting text from image

**Comment by matkoniecz** [comment]
Oh, you mean that it would also interpret it? Not only OCR?

I guess if on-device LLM would reliably work it would solve raised issues. (but I have no idea how much these on-device LLM can do and whether they are as on-device as advertised)

and additional problems are possible, do not rush to implement it

**Comment by mnalis** [comment]
There was some discussion about using LLM in https://github.com/streetcomplete/StreetComplete/issues/4775
